src/consumeroffset/kafka_offset_collect.go

```diff
@@ -74,39 +75,70 @@ func collectOffsetsForConsumerGroup(
 						clientPartitionWg.Done()
 						consumerGroupPartitionWg.Done()
 					}()
+					var partitionLag partitionLagResult
 					collectClientPartitionOffsetMetrics(
+						&partitionLag,
 						topicOffsetGetter,
 						consumerGroup,
 						description,
 						topic,
 						partition,
 						block,
-						clientPartitionLagChan,
-						cGroupPartitionLagChan,
 						kafkaIntegration,
 					)
+					if partitionLag.Topic == topic {
+						clientPartitionLagChan <- partitionLag
+						cGroupPartitionLagChan <- partitionLag
+					}
 				}(topic, partition, description, block)
 			}
 		}
 	}
 
 	if args.GlobalArgs.InactiveConsumerGroupOffset {
-		collectInactiveConsumerGroupOffsets(cGroupTopicLister, consumerGroup, topicExclussions, &consumerGroupPartitionWg, topicOffsetGetter, cGroupPartitionLagChan)
+		collectInactiveConsumerGroupOffsets(cGroupTopicLister, consumerGroup, topicExclusions, &consumerGroupPartitionWg, topicOffsetGetter, cGroupPartitionLagChan)
 	}
 
-	calculateClientLagTotals(clientPartitionLagChan, &clientPartitionWg, kafkaIntegration, consumerGroup)
-	calculateConsumerGroupLagTotals(cGroupPartitionLagChan, &consumerGroupPartitionWg, kafkaIntegration, consumerGroup)
+	clientMetricsAggregator := NewClientMetricsAggregator(consumerGroup)
+	lagResultsWg.Add(1)
+
+	go func() {
+		go func() {
+			clientPartitionWg.Wait()
+			log.Debug("Finished retrieving offsets for all member partitions in consumer group '%s'", consumerGroup)
+			close(clientPartitionLagChan)
+		}()
+		clientMetricsAggregator.waitAndAggregateMetrics(clientPartitionLagChan)
+		lagResultsWg.Done()
+	}()
+
+	cGroupMetricsAggregator := NewCGroupMetricsAggregator(consumerGroup, args.GlobalArgs.ConsumerGroupOffsetByTopic)
+	lagResultsWg.Add(1)
+
+	go func() {
+		go func() {
+			consumerGroupPartitionWg.Wait()
+			log.Debug("Finished retrieving offsets for all partitions in consumer group '%s'", consumerGroup)
+			close(cGroupPartitionLagChan)
+		}()
+		cGroupMetricsAggregator.waitAndAggregateMetrics(cGroupPartitionLagChan)
+		lagResultsWg.Done()
+	}()
+
+	lagResultsWg.Wait()
```

Comment:
If we define an aggregator interface this block of code can be common with a named function:

```go
func aggregateLagData(a aggregator, dataChan chan PartitionLagResult dataWg, aggregationWg sync.WaitGroup) {
    go func() { 
        // waits for dataWg and closes the dataChan 
    }()
    a.waitAndAggregateMetrics(dataChan)
    aggregationWg.Done()
}

// (...)

go aggregateLagData(clientMetricsAggregator, clientParitionLagChan, clientPartitionWg, lagResultsWg)
// (...)
go aggregateLagData(cGroupMetricsAggregator, cGroupPartitionLagChan, cGroupPartitionWg, lagResultsWg)

```

Reply:
Modified in last commit https://github.com/newrelic/nri-kafka/pull/172/commits/001a78f9b669832d9f7f44a3cfebc7d2fd9614e5